launch.go

```diff
@@ -293,53 +294,71 @@ func writeMetafile(metaSpace, metaFile, metaLog string, mergedMeta map[string]in
-		if pipelineID != parentJob.PipelineID {
-			// Write to "sd@123:component.json", where sd@123:component is the triggering job
-			externalMetaFile := "sd@" + strconv.Itoa(parentJob.PipelineID) + ":" + parentJob.Name + ".json"
-			writeMetafile(metaSpace, externalMetaFile, metaLog, parentBuild.Meta)
-			if join {
-				marshallValue, err := json.Marshal(parentBuild.Meta)
-				if err != nil {
-					return resultMeta, fmt.Errorf("Cloning meta of Parent Build ID %d: %v", parentBuildID, err)
-				}
-				var externalParentBuildMeta map[string]interface{}
-				json.Unmarshal(marshallValue, &externalParentBuildMeta)
+	// Sort by Endtime ASC
+	// Last finished parent build is priority
+	sort.SliceStable(parentBuilds, func(i, j int) bool {
+		return parentBuilds[i].Endtime.Before(parentBuilds[j].Endtime)
+	})
 
-				// Always exclude parameters from external meta
-				delete(externalParentBuildMeta, "parameters")
+	for _, parentBuild := range parentBuilds {
+		log.Printf("Fetching Parent Job %d", parentBuild.JobID)
+		parentJob, err := api.JobFromID(parentBuild.JobID)
+		if err != nil {
+			return resultMeta, fmt.Errorf("Fetching Job ID %d: %v", parentBuild.JobID, err)
+		}
 
-				resultMeta = deepMergeJSON(resultMeta, externalParentBuildMeta)
-			}
+		if parentBuild.Meta != nil {
+			// Check if build is from external pipeline
+			if pipelineID != parentJob.PipelineID {
+				// Write to "sd@123:component.json", where sd@123:component is the triggering job
+				externalMetaFile := "sd@" + strconv.Itoa(parentJob.PipelineID) + ":" + parentJob.Name + ".json"
+				writeMetafile(metaSpace, externalMetaFile, metaLog, parentBuild.Meta)
+				if join {
+					marshallValue, err := json.Marshal(parentBuild.Meta)
+					if err != nil {
+						return resultMeta, fmt.Errorf("Cloning meta of Parent Build ID %d: %v", parentBuild.ID, err)
+					}
+					var externalParentBuildMeta map[string]interface{}
+					json.Unmarshal(marshallValue, &externalParentBuildMeta)
+
+					// Always exclude parameters from external meta
+					delete(externalParentBuildMeta, "parameters")
 
-			// delete local version of external meta
-			pIDString := strconv.Itoa(parentJob.PipelineID)
-			pjn := parentJob.Name
-			if sdMeta, ok := resultMeta["sd"]; ok {
-				if externalPipelineMeta, ok := sdMeta.(map[string]interface{})[pIDString]; ok {
-					if _, ok := externalPipelineMeta.(map[string]interface{})[pjn]; ok {
-						delete(externalPipelineMeta.(map[string]interface{}), pjn)
+					resultMeta = deepMergeJSON(resultMeta, externalParentBuildMeta)
+				}
+
+				// delete local version of external meta
+				pIDString := strconv.Itoa(parentJob.PipelineID)
+				pjn := parentJob.Name
+				if sdMeta, ok := resultMeta["sd"]; ok {
+					if externalPipelineMeta, ok := sdMeta.(map[string]interface{})[pIDString]; ok {
+						if _, ok := externalPipelineMeta.(map[string]interface{})[pjn]; ok {
+							delete(externalPipelineMeta.(map[string]interface{}), pjn)
+						}
 					}
 				}
```

Comment:
please separate this logic to a different function, for example: 
```
func handleExternalPipelineMeta(api screwdriver.API, parentBuild screwdriver.Build, parentJob screwdriver.Job, resultMeta map[string]interface{}, metaSpace, metaLog string, isJoin bool) (map[string]interface{}, error) {
    externalMetaFile := ExternalMetaFilePrefix + strconv.Itoa(parentJob.PipelineID) + ":" + parentJob.Name + ExternalMetaFileSuffix
    writeMetafile(metaSpace, externalMetaFile, metaLog, parentBuild.Meta)

    if isJoin {
        marshallValue, err := json.Marshal(parentBuild.Meta)
        if err != nil {
            return resultMeta, fmt.Errorf("Cloning meta of Parent Build ID %d: %v", parentBuild.ID, err)
        }
        var externalParentBuildMeta map[string]interface{}
        json.Unmarshal(marshallValue, &externalParentBuildMeta)
        delete(externalParentBuildMeta, "parameters")
        resultMeta = deepMergeJSON(resultMeta, externalParentBuildMeta)
    }

    pIDString := strconv.Itoa(parentJob.PipelineID)
    pjn := parentJob.Name
    if sdMeta, ok := resultMeta["sd"]; ok {
        if externalPipelineMeta, ok := sdMeta.(map[string]interface{})[pIDString]; ok {
            if _, ok := externalPipelineMeta.(map[string]interface{})[pjn]; ok {
                delete(externalPipelineMeta.(map[string]interface{}), pjn)
            }
        }
    }

    return resultMeta, nil
}
```
this will help the code to be more readable.

Reply:
Thank you for suggestion. I fixed it 🙇